### PR TITLE
feat(atlas-signin): add atlas signin flow for the query bar AI input 

### PR DESCRIPTION
**Pull request by gribnoysup**
This is a proof on concept for the COMPASS-6954 to figure out what is exactly the scope of the work we will need to do to integrate atlas login through our oidc plugin. There is a bunch of TODOs here, I'm planning to base the design and required work tasks on this PR and after that will probably close this and split into multiple smaller parts. Opening as a draft so it's easier to keep track the required changes and discuss

Kapture 2023-06-30 at 19 41 04

### CODE REVIEW

File: packages/compass-query-bar/src/stores/ai-query-reducer.ts [hunk 1]
```diff
@@ -56,15 +74,17 @@ export type AIQuerySucceededAction = {
...
   Promise<void>,
   | AIQueryStartedAction
   | AIQueryCancelledAction
   | AIQueryFailedAction
   | AIQuerySucceededAction
 > => {
   return async (dispatch, getState) => {
-    const { aiQueryAbortController: existingAbortController } = getState();
+    const {
+      aiQuery: { aiQueryAbortController: existingAbortController },
```
Comment (Anemy) [inline packages/compass-query-bar/src/stores/ai-query-reducer.ts:86]: Nice catch with the thunk action not using the root 👍

Comment (gribnoysup) [reply packages/compass-query-bar/src/stores/ai-query-reducer.ts:86]: That's the only bug that I noticed, but there are other issues there like side-effects in reducer, non-serializable values in the store, using abort controller in connect directly, I'm assuming this is still poc (as is this PR anyway) and so haven't touched those. Are there any plans to address those, should I open some tickets?

File: packages/atlas-signin/src/main.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  private static calledOnce = false;
+
+  private static plugin = plugin.createMongoDBOIDCPlugin({
+    // TODO: as auth is done through the localhost loopback and not directly
+    // through customer okta, we will need to implement screens for following
+    // states: initial, success, error, not found; waiting for the designs
+    // redirectServerRequestHandler() {
+    // }
```
Comment (addaleax) [inline packages/atlas-signin/src/main.ts:20]: For this one, I’d suggest moving the `redirectServerRequestHandler` implementation that already exists in the `devtools-connect` package into its own package and then using that, I don’t think we really need different pages here right now.

(We may just want to move `devtools-connect` to the share monorepo, makes it even easier to do that split, how does that sound?)

File: packages/atlas-signin/src/main.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            this.token = token;
+            this.setupTokenRefresh();
+          });
+      },
+      Math.max(
+        // Min idle is 10mins
+        10 * 60 * 1000,
+        ((this.token?.expiresInSeconds ?? 0) * 1000) / 2
+      )
+    ).unref();
```
Comment (addaleax) [inline packages/atlas-signin/src/main.ts:95]: The oidc-plugin code already implements automatic refresh, I don’t think you need this function:

https://github.com/mongodb-js/oidc-plugin/blob/48ddba454c49a84b50a7b27b7513f6ce274109df/src/plugin.ts#L459-L470

File: packages/atlas-signin/src/main.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        return res.json();
+      }
+      const err = new Error(res.statusText);
+      (err as any).status = res.status;
+      throw err;
+    });
+  }
+
+  private static introspect() {
+    return fetch(`${this.issuer}/v1/introspect?client_id=${this.clientId}`, {
```
Comment (addaleax) [inline packages/atlas-signin/src/main.ts:124]: Fwiw, the OIDC client library we use also implements APIs for userinfo & token introspection: https://github.com/panva/node-openid-client/tree/main/docs#clientintrospecttoken-tokentypehint-extras

We could expose that from oidc-plugin if we want, I think.

File: packages/atlas-signin/src/main.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/atlas-signin/src/main.ts:20]: Sure, I can include this work in the project and tell design that we will not be implementing these new sign in flow states they apparently were working on. We will also need to modify the server not to serve a blank page on initial redirect and provide an additional handler for that

Comment (addaleax) [reply packages/atlas-signin/src/main.ts:20]: We were chatting about this during design sync on Thursday, and yes, we brought up that it probably makes sense to focus on designing the interfaces inside Compass itself. IIUC at least the actual auth flow pages in the customer Okta are not something we can design anyway?

File: packages/atlas-signin/src/main.ts (hunk 3 above)
Comment (gribnoysup) [reply packages/atlas-signin/src/main.ts:95]: Cool, good to know

File: packages/atlas-signin/src/main.ts (hunk 4 above)
Comment (gribnoysup) [reply packages/atlas-signin/src/main.ts:124]: Whatever you think makes more sense for oidc-plugin 👍

File: packages/atlas-signin/src/main.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/atlas-signin/src/main.ts:20]: I'm a bit lost now, sorry, so what was the result of the discussion: do we want the generic devtools connect screens for the login or special compass one that they seem to start working on? And yeah, we can't affect customer okta screens, but this I highlighted already some time ago in one of the figma discussions

Comment (addaleax) [reply packages/atlas-signin/src/main.ts:20]: Right, we want the generic screens (that already we use for DB OIDC login) for the initial iteration here :+1:

Comment (gribnoysup) [reply packages/atlas-signin/src/main.ts:20]: Got it, thanks for clarifying!

File: packages/compass-query-bar/src/stores/ai-query-reducer.ts (hunk 1 above)
Comment (Anemy) [reply packages/compass-query-bar/src/stores/ai-query-reducer.ts:86]: Addressed offline, updated in other prs.

**Comment by gribnoysup** [comment]
Closing in favour of https://github.com/mongodb-js/compass/pull/4661

**Pull request closed without merging**